### ISSUE TITLE
Discriminator gradient do not flow to few-step model?

### ISSUE THREAD

**Issue by ThibaultGROUEIX** [tianweiy/DMD2#75]
Hi @tianweiy ,

One more quick Q:
 During the generator turn, we need to collect two things:
1-  grads from the DMD loss (using no_grad() on dfake and dreal model to save memory/runtime)
2- grads from the gan i.e gan(dfake(fake_sample)). **For this we need to enable gradients to flow through dfake**

However, in the code, the gradients are disabled: 
https://github.com/tianweiy/DMD2/blob/main/main/sd_unified_model.py#L304-L310
```
                self.guidance_model.requires_grad_(False)
                loss_dict, log_dict = self.guidance_model(
                    generator_turn=True,
                    guidance_turn=False,
                    generator_data_dict=generator_data_dict
                )
                self.guidance_model.requires_grad_(True)
```
So I think, because of `self.guidance_model.requires_grad_(False)`, the GAN grads are not flowing through the generator.

Is my understanding correct?
Thanks
Thibault

**Comment by ThibaultGROUEIX** [comment]
Michael Gharbi answered this separately. 
`requires_grad_(False)` is not the same as `no_grad()` or `detach()`. The gradients are not stored for dfake but still flow through it to the generator.
